odoo/addons/base/models/ir_job.py

```diff
@@ -0,0 +1,156 @@
+
+    @classmethod
+    def _process_jobs(cls, db, jobs, manager_cr):
+        # @juc method too long ? split call/error/success ?
+        for job in jobs:
+            job_ids = json.loads(job['ids'])
+            job_args = json.loads(job['args'])
+            job_kwargs = json.loads(job['kwargs'])
+
+            with db.cursor() as job_cr, api.Environment.manage():
+                env = api.Environment(job_cr, job['user_id'])
+                # @juc we could put a special key in the env like "async_job": True, would it be necessary ?
+                try:
+                    records = env[job['model']].browse(json.loads(job_ids))
+                    res = getattr(records, job['method'])(*job_args, **job_kwargs)
+                    job_result = json.dumps(res)
+                    records.flush()
+                except psycopg2.OperationalError:
+                    job_scheduled = fields.Datetime.to_datetime(job['scheduled'])
+                    later = fields.Datetime.now() + timedelta(milliseconds=1000 + randint(0, 100))
+                    manager_cr.execute("""
+                        UPDATE ir_job
+                        SET scheduled = %s
+                        WHERE id = %d
+                        """, (fields.Datetime.to_string(later), job['id']))
+                except Exception:
+                    header, n, tail_tb = traceback.format_exc().partition('\n')
+                    job_traceback = header + n + job['traceback'] + tail_tb
+                    # @juc the tail_tb will contain _process_jobs and up, we could strip those 'up' lines
+                    # @juc log even more ? like ids/args/kwargs ?
+                    _logger.error('Failed to process async job:\n%s', job_traceback)
+                    manager_cr.execute("""
+                        UPDATE ir_job
+                        SET state = 'failed'
+                            traceback = %s
+                        WHERE id = %d
+                        """, (job_traceback, job['id']))
+                    if job['bus_channel']:
+                        self.env['bus.bus'].sendone(job['bus_channel'], {
+                            'state': 'failure',
+                            'traceback': job_traceback,
+                        })
+                else:
+                    manager_cr.execute("""
+                        UPDATE ir_job
+                        SET state = 'done'
+                            result = %s
+                        WHERE id = %d
+                        """, (job_result, job['id']))
+                    if job['bus_channel']:
+                        self.env['bus.bus'].sendone(job['bus_channel'], {
+                            'state': 'success',
+                            'result': res
+                        })
+
+
+    @staticmethod
+    def _sleep_time(job_cnt, daemon_no):
+        return 0 if job_cnt else .1 + (daemon_no % 10) / 100
+
+# async/call_soon/sched
+def async(self, method, *args, **kwargs):
```

Comment:
just fyi, `async` is a reserved keyword in Python >= 3.7

Reply:
ikr